[CPU] Added torch.bmm for complex tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42383 [CPU] Added torch.bmm for complex tensors**

Test Plan - Updated existing tests to run for complex dtypes as well. 

Also added tests for `torch.addmm`, `torch.badmm`

Differential Revision: [D22960339](https://our.internmc.facebook.com/intern/diff/D22960339)

anjali411 committed:
[CPU] Added torch.bmm for complex tensors

[ghstack-poisoned]

zasdfgbnm commented on test/test_torch.py:
```diff
@@ -17135,7 +17139,7 @@ def test_lcm(self, device, dtype):
+    @dtypesIfCUDA(torch.float32, torch.float64, torch.cfloat, torch.cdouble)
     @tf32_on_and_off(0.01)
     def test_mm(self, device, dtype):
         def _test_mm(n, m, p, dtype, genf):
```
LOL, GitHub UI is so confusing.

<img width="492" alt="Screenshot_20200731_131828" src="https://user-images.githubusercontent.com/1032377/89074055-7c912780-d330-11ea-9304-f9ef9502a4e3.png">

zasdfgbnm commented on test/test_torch.py:
```diff
@@ -19438,23 +19442,23 @@ def inner(self, device, dtype):
         1e-1, 1e-1, 1e-5, _float_types2),
     ('addbmm', '', _small_2d, lambda t, d: [_small_3d(t, d), _small_3d(t, d)],
-        1e-1, 1e-1, 1e-4, _float_types2, _cpu_types, True, [tf32_on_and_off(0.005)]),
+        1e-1, 1e-1, 1e-4, _complex_and_float_types2, _cpu_types + _complex_types, True, [tf32_on_and_off(0.005)]),
```
@mruberry IIUC, `_cpu_types` is empty list, and `_cpu_types + _complex_types` would mean run CPU test only with complex?

mruberry replied:
Boy does this set of tests need to be entirely rewritten. 

Your understanding is correct. Note that "testing" on CPU here just means verifying that the function and method variants of the op produce the same value. There's no check those values are correct.

anjali411 replied:
I didn't know that the cpu tests here only compare the function and method result which basically doesn't test anything for a lot of ops. will remove it.

ngimel replied:
"testing" on cpu means results in low precision (bfloat16 or whatever _cpu_types is) is compared to result in fp32 precision. For _cpu_types=fp32 it would indeed just compare function and method.

anjali411 committed:
Update on "[CPU] Added torch.bmm for complex tensors"


Test Plan - Updated existing tests to run for complex dtypes as well. 

Also added tests for `torch.addmm`, `torch.badmm`

[ghstack-poisoned]

mruberry commented on test/test_torch.py:
```diff
@@ -14292,6 +14293,9 @@ def test_blas_nan_out(self, device):
         np_out = torch.full((n, p), float('nan'), device=device)
         self.assertEqual(torch.mm(nm, mp), torch.mm(nm, mp, out=np_out))
 
+        if dtype.is_complex and device.startswith('cuda'):
```
`self.device_type == 'cuda'` or `torch.device(device).device_type`

zasdfgbnm commented on test/test_torch.py:
```diff
@@ -17229,7 +17233,7 @@ def genf_float(x, y):
 
     @onlyCPU
-    @dtypes(torch.float)
+    @dtypes(*(torch.testing.get_all_complex_dtypes() + [torch.float, torch.double]))
```
NIT: Could we unify

```
torch.float32, torch.float64, torch.cfloat, torch.cdouble
```

vs

```
torch.testing.get_all_complex_dtypes() + [torch.float, torch.double]
```

anjali411 replied:
do you mean, replace all instances of `torch.float32, torch.float64, torch.cfloat, torch.cdouble` with `torch.testing.get_all_complex_dtypes() + [torch.float, torch.double]` ?

zasdfgbnm replied:
I mean the opposite:
```
torch.float32, torch.float64, torch.cfloat, torch.cdouble
```
looks better than
```
*(torch.testing.get_all_complex_dtypes() + [torch.float, torch.double])
```

anjali411 replied:
I think it's better to use `torch.testing.get_all_complex_dtypes()` since in case we add more complex dtypes in future (for example, support for torch.chalf or others), we wouldn't have to update the tests one by one to add tests for that complex dtype. Just including it in `torch.testing.get_all_complex_dtypes()` would suffice.

zasdfgbnm reviewed (approved):
Overall LGTM

zasdfgbnm commented:
I think you need to rebase to fix the docker image not found error.

anjali411 committed:
Update on "[CPU] Added torch.bmm for complex tensors"


Test Plan - Updated existing tests to run for complex dtypes as well. 

Also added tests for `torch.addmm`, `torch.badmm`

[ghstack-poisoned]

anjali411 committed:
Update on "[CPU] Added torch.bmm for complex tensors"


Test Plan - Updated existing tests to run for complex dtypes as well. 

Also added tests for `torch.addmm`, `torch.badmm`

Differential Revision: [D22960339](https://our.internmc.facebook.com/intern/diff/D22960339)

[ghstack-poisoned]

anjali411 committed:
Update on "[CPU] Added torch.bmm for complex tensors"


Test Plan - Updated existing tests to run for complex dtypes as well. 

Also added tests for `torch.addmm`, `torch.badmm`

Differential Revision: [D22960339](https://our.internmc.facebook.com/intern/diff/D22960339)

[ghstack-poisoned]

ngimel commented:
@jeffdaily currently cgemm/zgemm is disabled on [rocm](https://github.com/pytorch/pytorch/blob/faca3c43e6b7d6e97440608898956773dbe2c667/aten/src/ATen/cuda/CUDABlas.cpp#L186-L200), but rocblas actually supports it - what would it take to enable it?
